### PR TITLE
DoubleEndedIterator for Args

### PR DESCRIPTION
**Pull request by Byron**
This PR implements the DoubleEndedIterator trait for the `std::env::Args[Os]` structure, as well
as the internal implementations.

It is primarily motivated by me, as I happened to implement a simple `reversor` program many times
now, which so far had to use code like this:

``` Rust
for arg in std::env::args().skip(1).collect::<Vec<_>>().iter().rev() {}
```

... even though I would have loved to do this instead:

``` Rust
for arg in std::env::args().skip(1).rev() {}
```

The latter is more natural, and I did not find a reason for not implementing it.
After all, on every system, the number of arguments passed to the program are known
at runtime.
### About the implementation

To my mind, it follows KISS, and does not try to be smart at all. Also, there are no unit-tests,
primarily as I did not find any existing tests for the `Args` struct either.
### Critical Comments

The windows implementation is basically a copy-pasted variant of the `next()` method implementation,
and I could imagine sharing most of the code instead. Actually I would be happy if the reviewer would
ask for it.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Looks good to me, thanks for the PR! As you mentioned, can you share code in the Windows implementation to deduplicate, and then can you also add a few tests for this?

**Comment by Byron** [comment]
Thanks @alexcrichton for the hints ! Currently I am trying to see how to best run the tests. For now, the fasted run I got was via `time make check-stage1-std`, which took just 10.5m . I also tried `time make check-stage1-std TESTNAME=env::tests`, but that one takes even longer.

Do you know how to speed it up testing in my case ?

**Comment by Byron** [comment]
Never mind - I ran `make check-stage1-std -n` to see the commands actually executed, and from there I could cut down the time for each run to about 2 minutes. That's as good as it gets, as the compilation took 1:44 minutes already.

**Comment by alexcrichton** [comment]
Another option may be the `NO_REBUILD` environment variable which should avoid recompiling the whole of libstd and instead go straight to tests. Finally you may wish to pass `./configure --disable-optimize-tests` as that helps stdtest compile a little more quickly

File: src/libstd/env.rs [hunk 1]
```diff
@@ -588,6 +588,13 @@ impl ExactSizeIterator for Args {
 }
 
 #[stable(feature = "env", since = "1.0.0")]
+impl DoubleEndedIterator for Args {
+    fn next_back(&mut self) -> Option<String> {
+        self.inner.next_back().map(|s| s.into_string().unwrap())
+    }
+}
+
+#[stable(feature = "env", since = "1.0.0")]
```
Comment (sfackler) [inline src/libstd/env.rs]: This should be marked `since = "1.11.0"`, not `1.0.0`

File: src/libstd/env.rs [hunk 2]
```diff
@@ -599,6 +606,10 @@ impl ExactSizeIterator for ArgsOs {
     fn len(&self) -> usize { self.inner.len() }
 }
 
+#[stable(feature = "env", since = "1.0.0")]
```
Comment (sfackler) [inline src/libstd/env.rs]: ^

**Comment by sfackler** [comment]
Can you add on `ExactSizeIterator` impls while we're at it?

**Comment by Byron** [comment]
@sfackler Good idea, ExactSizeIterator shall be done as well !

**Comment by Byron** [comment]
@sfackler When running `make tidy` with the requested changes to `since`, I get an error saying:
```
* 421 error codes
* highest error code: E0524
/Users/byron/Documents/dev/rust/rust-lang/rust/src/libstd/env.rs:590: different `since` than before
/Users/byron/Documents/dev/rust/rust-lang/rust/src/libstd/env.rs:609: different `since` than before
thread '<main>' panicked at 'some tidy checks failed', /Users/byron/Documents/dev/rust/rust-lang/rust/src/tools/tidy/src/main.rs:51
note: Run with `RUST_BACKTRACE=1` for a backtrace.
make: *** [tidy] Error 101
```

I assume that I have to create a new feature and assign it a since 1.11 accordingly. If so, what is the feature name you would like to see here ?
Thank you

**Comment by Byron** [comment]
I think this PR is ready for review. Please note that I am not sure my feature-name is desirable.

Also I found ExactSizeIterator to be implemented already.

File: src/libstd/env.rs [hunk 3]
```diff
@@ -928,6 +939,16 @@ mod tests {
...
+    
+    #[test]
+    fn test_env_args_reverse_iterator() {
+        let args: Vec<_> = args().rev().collect();
+        let l = args.len();
+        assert!(l >= 3);
+        assert!(args[l-1].contains("stdtest"));
+        assert_eq!(args[1], "--logfile");
+        assert!(args[0].ends_with(".log"));
+    }
```
Comment (alexcrichton) [inline src/libstd/env.rs]: Could you add a unit test elsewhere with this test as this'd unfortunately break running the binary manually :(

Comment (alexcrichton) [inline src/libstd/env.rs]: (e.g., adding a test to src/test/run-pass where the test just executes itself to run the test)

**Comment by alexcrichton** [comment]
Ah yeah the test name is fine, we tend to just pick totally arbitrary names anyway :)

**Comment by Byron** [comment]
Thanks for the feedback ! The problem I see with the run-pass tests is the inability to pass arguments to the program, making it impossible to be sure the feature is implemented correctly. Do you have an idea who that could be done ?

**Comment by alexcrichton** [comment]
Ah yeah the trick I normally take looks like [this test](https://github.com/rust-lang/rust/blob/master/src/test/run-pass/command-exec.rs) where a test re-executes itself to run the test and then asserts a property of the output

**Comment by Byron** [comment]
Ok, it seems to be done :) !

A few things I noticed:
* the run-pass test does not need to specify a feature. Apparently trait implementations are not affected by feature gates ?
* I am unsure if the feature name `env_iterators` is descriptive enough, and I'd be happy to hear other suggestions.

File: src/libstd/lib.rs [hunk 4]
```diff
@@ -271,6 +271,7 @@
 #![feature(vec_push_all)]
 #![feature(zero_one)]
 #![feature(question_mark)]
+#![feature(env_iterators)]
```
Comment (alexcrichton) [inline src/libstd/lib.rs]: Ah this isn't actually required for stable features, so it's ok to leave this off

**Comment by alexcrichton** [comment]
Thanks! Yeah feature annotations aren't required for stable features (which this is), and we generally pick totally arbitrary names for these so `env_iterators` is fine :)

**Comment by Byron** [comment]
Alright, I have removed the feature, which indeed is not required. The test was made a little less-trusty too.
Let's see what travis thinks about this.

**Comment by alexcrichton** [comment]
That all looks good to me, thanks! Could you squash down into one commit as well?

**Comment by Byron** [comment]
Alright, it's done :) !

**Comment by Byron** [comment]
I believe to have fixed the function signature to not cause build failures on windows anymore.

**Comment by Byron** [comment]
Alright, getting there ! With a little bit of luck, this totally works now.

**Comment by Byron** [comment]
@alexcrichton Could you have a look again please ? It should go through on windows now. Thank you.

**Comment by Byron** [comment]
@alexcrichton The failure seems to be unrelated:
```
Synchronizing submodule url for 'src/rt/hoedown'
Synchronizing submodule url for 'src/rust-installer'
fatal: Needed a single revision
Unable to find current revision in submodule path 'src/llvm'
```
Is there anything I can do? Thanks.

**Comment by alexcrichton** [comment]
@bors: retry

nah we just have a lot of network failures recently :(

**Comment by Byron** [comment]
@alexcrichton I have placed the unsafe scope into the right place, and rebased everything into one commit which now will hopefully go through. My apologies for all the attempts it took to make work this simple patch.

**Comment by alexcrichton** [comment]
@Byron no worries! Most failures nowadays are due to our infrastructure anyways :)

@bors: r+ 1aa8dad854155221db7cec19b6105c673e4a871e

**Pull request merged**